Added well-known OpenId auto configuration to the SSO generic-provider

gbarbon opened the pull request:
Added automatic retrieval of the required OpenID URIs through the use of the `well-known` OpenID configuration document (see [here](https://openid.net/specs/openid-connect-discovery-1_0.html) for further information).

**Related Issue**
This PR fixes/closes _#3044_

**Description of the solution adopted**
The SSO generic-provider now reads the `well-known` configuration document provided by the OpenID Provider and configures the URIs consequently. If it is not able to configure one of the URIs using the `well-known`, it tries to read the related property (provided by the user).

**Screenshots**
_N/A_

**Any side note on the changes made**
The `sso.md` dcoumentation file has been updated accordingly.

gbarbon committed:
Use the well-known configuration document to configure the SSO OpenID providers

Signed-off-by: Gianluca Barbon <gianluca.barbon@eurotech.com>

gbarbon committed:
Simplifying well-known conf path generation

Signed-off-by: Gianluca Barbon <gianluca.barbon@eurotech.com>

gbarbon committed:
Improved documentation concerning clientId and audience

Signed-off-by: Gianluca Barbon <gianluca.barbon@eurotech.com>

LeoNerdoG commented on sso/provider/src/main/java/org/eclipse/kapua/sso/provider/SingleSignOnUtils.java:
```diff
@@ -0,0 +1,112 @@
+/*******************************************************************************
+ * Copyright (c) 2019, 2020 Eurotech and/or its affiliates and others
```
hey @gbarbon, if this file is new, header should contain only year "2020" and not 2019.

gbarbon replied:
Yep, that's true! I think this happened because I renamed the old `JsonUtils.java` class to `SingleSignOnUtils`. Anyway, now it's fixed.

Coduz commented on sso/provider-generic/src/main/java/org/eclipse/kapua/sso/provider/generic/GenericSingleSignOnService.java:
```diff
@@ -34,29 +45,70 @@ public GenericSingleSignOnService(final SsoSetting ssoSettings, final GenericSso
+            }
+        } catch (SsoException se) {
+            String authUri = genericSettings.getString(GenericSsoSettingKeys.SSO_OPENID_SERVER_ENDPOINT_AUTH);
+            if (authUri == null || authUri.isEmpty()) {
```
Please make use of `Strings.isNullOrEmpty`

Coduz replied:
Please make use of `Strings.isNullOrEmpty`

Coduz commented on sso/provider-generic/src/main/java/org/eclipse/kapua/sso/provider/generic/GenericSingleSignOnService.java:
```diff
@@ -34,29 +45,70 @@ public GenericSingleSignOnService(final SsoSetting ssoSettings, final GenericSso
+            }
+        } catch (SsoException se) {
+            String tokenUri = genericSettings.getString(GenericSsoSettingKeys.SSO_OPENID_SERVER_ENDPOINT_TOKEN);
+            if (tokenUri == null || tokenUri.isEmpty()) {
```
Please make use of `Strings.isNullOrEmpty`

Coduz commented on sso/provider-generic/src/main/java/org/eclipse/kapua/sso/provider/generic/GenericSingleSignOnService.java:
```diff
@@ -34,29 +45,70 @@ public GenericSingleSignOnService(final SsoSetting ssoSettings, final GenericSso
+            }
+        } catch (SsoException se) {
+            String logoutUri = genericSettings.getString(GenericSsoSettingKeys.SSO_OPENID_SERVER_ENDPOINT_LOGOUT);
+            if (logoutUri == null || logoutUri.isEmpty()) {
```
Please make use of `Strings.isNullOrEmpty`

Coduz commented on sso/provider-generic/src/main/java/org/eclipse/kapua/sso/provider/generic/GenericSingleSignOnService.java:
```diff
@@ -34,29 +45,70 @@ public GenericSingleSignOnService(final SsoSetting ssoSettings, final GenericSso
+     */
+    private String getOpenIdConfPath() throws SsoIllegalArgumentException {
+        String issuerUri = genericSettings.getString(GenericSsoSettingKeys.SSO_OPENID_JWT_ISSUER_ALLOWED);
+        if (issuerUri == null || issuerUri.isEmpty()) {
```
Please make use of `Strings.isNullOrEmpty`

Coduz commented on sso/provider/src/main/java/org/eclipse/kapua/sso/provider/SingleSignOnUtils.java:
```diff
@@ -0,0 +1,112 @@
+     */
+    public static String getOpenIdConfPath(String issuer) throws SsoIllegalArgumentException {
+        String openIDConfPathSuffix = SsoSetting.getInstance().getString(SsoSettingKeys.SSO_OPENID_CONF_PATH, DEFAULT_SSO_OPENID_CONF_PATH);
+        if (openIDConfPathSuffix == null || openIDConfPathSuffix.isEmpty()) {
```
Please make use of `Strings.isNullOrEmpty`

Coduz commented on sso/provider-generic/src/main/java/org/eclipse/kapua/sso/provider/generic/GenericSingleSignOnService.java:
```diff
@@ -34,29 +45,70 @@ public GenericSingleSignOnService(final SsoSetting ssoSettings, final GenericSso
+    protected String getLogoutUri() throws SsoException {
+        try {
+            final Optional<URI> uri = SingleSignOnUtils.getConfigUri(LOGOUT_WELL_KNOWN_KEY, getOpenIdConfPath());
+            if (uri.isPresent()) {
```
Please leverage the `Optional` capabilities.
i.e.
```
Optional<Object> optional = Optional.ofNullable({somevalue});
Object object = optional.orElseThrow(Exception::new);
```

Coduz commented on sso/provider-generic/src/main/java/org/eclipse/kapua/sso/provider/generic/GenericSingleSignOnService.java:
```diff
@@ -34,29 +45,70 @@ public GenericSingleSignOnService(final SsoSetting ssoSettings, final GenericSso
+    protected String getAuthUri() throws SsoException {
+        try {
+            final Optional<URI> uri = SingleSignOnUtils.getConfigUri(AUTH_WELL_KNOWN_KEY, getOpenIdConfPath());
+            if (uri.isPresent()) {
```
Please leverage the `Optional` capabilities.
i.e.
```
Optional<Object> optional = Optional.ofNullable({somevalue});
Object object = optional.orElseThrow(Exception::new);
```

Coduz reviewed (changes requested):
Hi @gbarbon ,

I've made just small comments on usage of `Optional` and the String blank checking.

_Alberto_

gbarbon committed:
Using Optional class capabilities, Strings.isNullOrEmpty and CollectionUtils.isEmpty

Signed-off-by: Gianluca Barbon <gianluca.barbon@eurotech.com>

gbarbon commented:
> Hi @gbarbon ,
> 
> I've made just small comments on usage of `Optional` and the String blank checking.
> 
> _Alberto_

Thanks @Coduz , I implemented everything as suggested, and also added CollectionUtils.isEmpty to check for empty List
